core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java

```diff
@@ -427,6 +428,24 @@ public ProxiedAddresses proxiedAddresses() {
     }
 
     @Override
+    public CompletableFuture<Void> initiateConnectionShutdown() {
+        return initiateConnectionShutdown(null);
+    }
+
+    @Override
+    public CompletableFuture<Void> initiateConnectionShutdown(@Nullable Duration gracePeriod) {
```

Comment:
Could you also add an overloaded method that accepts `long gracePeriodMillis` where `-1` means 'use the default grace period' and `0` means 'no grace period'? Armeria almost always provides overloaded methods when accepting a `Duration`. For performance benefits (less objects to juggle with), we use `long millis` or `long nanos` field internally as well, so a method that accepts a `Duration` is usually a shortcut that calls the method that accepts a `long`.

Reply:
See https://github.com/line/armeria/pull/3715/files#r676902455 for my concern using negative value as a default.

Is there a good reason to worry about performance in this case? These objects are configuration and not created / destructed frequently, it should have negligible impact on performance.

Reply:
> Is there a good reason to worry about performance in this case? These objects are configuration and not on the critical path, not created / destructed frequently, they should have negligible impact on performance and memory usage.

It's probably my habit that tries to keep everything lean, because everything adds up eventually in a framework. In a framework, we don't really know for sure how our users will use this. For example, a user might calculate the grace period dynamically based on some factor, then suddenly it will be allocating many objects than usual, then the user will see the increased GC pressure at certain cases. So the rule of thumb is to make sure we optimize a little bit more aggressively than when we write service code.